http/etag.ts

```diff
@@ -0,0 +1,196 @@
+
+  /** Override the default behavior of calculating the `ETag`, either forcing
+   * a tag to be labelled weak or not. */
+  weak?: boolean;
+}
+
+function isFileInfo(value: unknown): value is FileInfo {
+  return Boolean(
+    value && typeof value === "object" && "mtime" in value && "size" in value,
+  );
+}
+
+async function calcEntity(
+  entity: string | Uint8Array,
+  { algorithm }: ETagOptions,
+) {
+  // a short circuit for zero length entities
+  if (!entity.length && !algorithm) {
+    return "811c9dc5";
+  }
+
+  return toHashString(await createHash(algorithm ?? DEFAULT_ALGORITHM, entity));
+}
+
+async function calcFileInfo(fileInfo: FileInfo, { algorithm }: ETagOptions) {
+  if (fileInfo.mtime) {
+    return toHashString(
+      await createHash(
+        algorithm ?? DEFAULT_ALGORITHM,
+        `${fileInfo.mtime.toJSON()}${fileInfo.size}`,
+      ),
+    );
+  }
+  return HASHED_DENO_DEPLOYMENT_ID;
+}
+
+/** Calculate an ETag for an entity. When the entity is a specific set of data
+ * it will be fingerprinted as a "strong" tag, otherwise if it is just file
+ * information, it will be calculated as a weak tag.
+ *
+ * ```ts
+ * import { calculate } from "https://deno.land/std@$STD_VERSION/http/etag.ts";
+ * import { assert } from "https://deno.land/std@$STD_VERSION/_util/asserts.ts"
+ *
+ * const body = "hello deno!";
+ *
+ * const etag = await calculate(body);
+ * assert(etag);
+ *
+ * const res = new Response(body, { headers: { etag } });
+ * ```
+ */
+export async function calculate(
+  entity: Entity,
+  options: ETagOptions = {},
+): Promise<string | undefined> {
+  const weak = options.weak ?? isFileInfo(entity);
+  const tag =
+    await (isFileInfo(entity)
+      ? calcFileInfo(entity, options)
+      : calcEntity(entity, options));
+
+  return tag ? weak ? `W/"${tag}"` : `"${tag}"` : undefined;
+}
+
+/** A helper function that takes the value from the `If-Match` header and a
+ * calculated etag for the target. By using strong comparison, return `true` if
+ * the values match, otherwise `false`.
+ *
+ * See MDN's [`If-Match`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/If-Match)
+ * article for more information on how to use this function.
+ *
+ * ```ts
+ * import {
+ *   calculate,
+ *   ifMatch,
+ * } from "https://deno.land/std@$STD_VERSION/http/etag.ts";
+ * import { serve } from "https://deno.land/std@$STD_VERSION/http/server.ts";
+ * import { assert } from "https://deno.land/std@$STD_VERSION/_util/asserts.ts"
+ *
+ * const body = "hello deno!";
+ *
+ * await serve(async (req) => {
+ *   const ifMatchValue = req.headers.get("if-match");
+ *   const etag = await calculate(body);
+ *   assert(etag);
+ *   if (!ifMatchValue || ifMatch(ifMatchValue, etag)) {
+ *     return new Response(body, { status: 200, headers: { etag } });
+ *   } else {
+ *     return new Response(null, { status: 412, statusText: "Precondition Failed"});
+ *   }
+ * });
+ * ```
+ */
+export function ifMatch(
```

Comment:
Imo the name is a bit misleading. In the js landscape, `match` (in String, RegExp etc.) normally returns a matched content. I think `test` would make more sense as a name as it returns a boolean.

Reply:
They align to the header names, ergo align to the logic the header represents.